### PR TITLE
kvserver: communicate snapshot index back along with snapshot response

### LINKED ISSUE #84242: kv/kvserver: TestAdminRelocateRange failed
kv/kvserver.TestAdminRelocateRange [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5720159?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5720159?buildTab=artifacts#/) on master @ [571bfa3afb3858ae84d8a8fcdbb0a38e058402a5](https://github.com/cockroachdb/cockroach/commits/571bfa3afb3858ae84d8a8fcdbb0a38e058402a5):

```
=== RUN   TestAdminRelocateRange
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/751d67000aac5f3394c2369309253f02/logTestAdminRelocateRange2871936122
    test_log_scope.go:80: use -show-logs to present logs inline
    client_relocate_range_test.go:238: 
        	Error Trace:	client_relocate_range_test.go:189
        	            				client_relocate_range_test.go:238
        	Error:      	Not equal: 
        	            	expected: 3
        	            	actual  : 2
        	Test:       	TestAdminRelocateRange
        	Messages:   	wrong number of atomic changes
    client_relocate_range_test.go:238: all changes:
    client_relocate_range_test.go:238: 1: [{ADD_VOTER n4,s4} {REMOVE_VOTER n1,s1}]
    client_relocate_range_test.go:238: 2: [{ADD_VOTER n6,s6} {REMOVE_VOTER n3,s3}]
    panic.go:661: -- test log scope end --
--- FAIL: TestAdminRelocateRange (1.93s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestAdminRelocateRange.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

Jira issue: CRDB-17542

### LINKED ISSUE #97971: kvserver: avoid unnecessary raft snapshots
Fix the reasons for unnecessary snapshots and land https://github.com/cockroachdb/cockroach/pull/87554. See that PR for discussion.

> @tbg I wanted to get this test pushed because it will be useful to have this fixed to handle some of the more strict tests related to when we need snapshots. While I understand there are cases where Raft Snapshots are sent unnecessarily, reducing these cases will help allow the writing of less flaky and more strict tests. Since you and @pavelkalinnikov are cleaning up some code related to Raft, having this test might be a useful thing to fix.

_Originally posted by @andrewbaptist in https://github.com/cockroachdb/cockroach/issues/87554#issuecomment-1402385615_
            

Jira issue: CRDB-24995

### PR DESCRIPTION
**Pull request by tbg**
This addresses the following race:

- n1 runs a ConfChange that adds n2 as a learner.
- n1 sends MsgApp to the learner.
- n1 starts the INITIAL snapshot, say at index 100.
- n2 receives n1's MsgApp. Since it's an uninitialized
  Replica and its log is empty, it rejects this MsgApp.
- n2 receives and applies the INITIAL snapshot, which prompts it to send an
  affirmative MsgAppResp to n1.
- n1's RawNode now tracks n2 as StateProbe (via call to ReportSnapshot(success))
- n1 receives the MsgApp rejection; n2 regresses to StateSnapshot because the
  rejection comes with a RejectHint (suggested next index to try) of zero,
  which is not in n1's log. In particular, the SnapshotIndex will likely be
  higher than the index of the snapshot actually sent, say 101.
- n1 receives the affirmative MsgAppResp (for index 100). However, 100 < 101
  so this is ignored and the follower remains in StateSnapshot.

With this commit, the last two steps cannot happen: n2 transitions straight to
StateReplicate because we step a copy of the affirmative MsgAppResp in. The
later rejection will be dropped, since it is stale (you can't hint at index zero
when you already have a positive index confirmed).

I will add that there is no great testing for the above other than stressing
the test with additional logging, noting the symptoms, and noting that they
disappear with this commit. Scripted testing of this code is within reach[^1]
but is outside of the scope of this PR.

[^1]: https://github.com/cockroachdb/cockroach/issues/105177

There is an additional bit of brittleness that is silently suppressed by this
commit, but which deserves to be fixed independently because how the problem
gets avoided seems accidental and incomplete. When raft requests a snapshot, it notes its
current LastIndex and uses it as the PendingSnapshot for the follower's
Progress.

At the time of writing, MsgAppResp that reconnect the follower to the log but
which are not greater than or equal to PendingSnapshot are ignored. In effect,
this means that perfectly good snapshots are thrown away if they happen to be a
little bit stale. In the example above, the snapshot is stale: PendingSnapshot
is 101, but the snapshot is at index 100. Then how does this commit (mostly)
fix the problem, i.e. why isn't the snapshot discarded? The key is that when we
synchronously step the MsgAppResp(100) into the leader's RawNode, the rejection
hasn't arrived yet and so the follower transitions into StateReplicate with a
Match of 100. This is then enough so that raft recognizes the rejected MsgApp
as stale (since it would regress on durably stored entries). However, there is
an alternative example where the rejection arrives earlier: after the snapshot
index has been picked, but before the follower has been transitioned into
StateReplicate. For this to have a negative effect, an entry has to be appended
to the leader's log between generating the snapshot and handling the rejection.
Without the combination of delegated snapshots and sustained write activity on
the leader, this window is small, and this combination is usually not present
in tests but it may well be relevant in "real" clusters. We track addressing
this in #106813.

Closes #87554.
Closes #97971.
Closes #84242.

Epic: None
Release note (bug fix): removed a source of unnecessary Raft snapshots during
replica movement.

### CODE REVIEW

File: pkg/kv/kvserver/replica_learner_test.go [hunk 1]
```diff
@@ -2441,3 +2441,53 @@ func TestRebalancingAndCrossRegionZoneSnapshotMetrics(t *testing.T) {
...
+				Store: &kvserver.StoreTestingKnobs{DisableRaftSnapshotQueue: true}},
+			},
+			ReplicationMode: base.ReplicationManual,
+		},
+	)
+	defer tc.Stopper().Stop(ctx)
+
+	key := tc.ScratchRange(t)
+	tc.AddVotersOrFatal(t, key, tc.Target(1))
+	// This may hang since without the snapshot queue, the raft group is broken.
```
Comment (erikgrinaker) [inline pkg/kv/kvserver/replica_learner_test.go:2491]: nit: consider mentioning that the operation has an internal timeout, and won't hang "forever".

File: pkg/kv/kvserver/replica_raft.go [hunk 2]
```diff
@@ -951,6 +951,17 @@ func (r *Replica) handleRaftReadyRaftMuLocked(
 			if err := r.applySnapshot(ctx, inSnap, snap, hs, subsumedRepls); err != nil {
 				return stats, errors.Wrap(err, "while applying snapshot")
 			}
+			for _, msg := range msgStorageAppend.Responses {
+				// TODO is this conditional enough?
```
Comment (erikgrinaker) [inline pkg/kv/kvserver/replica_raft.go:955]: I think so? At least naïvely, it seems to be.

**Review by erikgrinaker — APPROVED** [review]

File: pkg/kv/kvserver/raft_transport.go [hunk 3]
```diff
@@ -1132,18 +1132,18 @@ func (t *RaftTransport) SendSnapshot(
 // and determines if it encountered any errors when sending the snapshot.
 func (t *RaftTransport) DelegateSnapshot(
 	ctx context.Context, req *kvserverpb.DelegateSendSnapshotRequest,
-) error {
+) (*kvserverpb.DelegateSnapshotResponse, error) {
```
[the remark this answers is not in the document]
Comment (tbg) [reply pkg/kv/kvserver/raft_transport.go:1135]: I went the other way and am returning the snapshot response instead consistently.

**Comment by tbg** [comment]
TestSqlActivityUpdateTopLimitJob flaked in [CI](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_UnitTests_BazelUnitTests/11008238?buildTab=overview&hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true&expandBuildProblemsSection=true&expandBuildTestsSection=true#%2Fbazel-testlogs%2Fpkg%2Fsql%2Fsql_test%2Fshard_16_of_16;%2Ftmp%2F_tmp%2F9f172280e0add88a3f8251875cbae1e5%2FlogTestSqlActivityUpdateTopLimitJob604373916) with a package timeout. However, the test itself ran only for 15s, so it was something else in that package that took it over the 14m55s threshold.

**Comment by tbg** [comment]
bors r=erikgrinaker
TFTR!

**Pull request merged**